File: Backtracking/Python/MaximalString.py
class Solution:
    # @param A : string
    # @param B : integer
    # @return a strings
    #backtrack naciagany na siłe, w cpp dla ponizszych danych program sie wywala
    # a tutaj iteracyjnie wszystko git, ale nie zalicza rozwiązania
    #10343456789765432457689065543876
    #10
    # w c++ backtracking z komentarza jest uznawany, choć dla długości A ~kilkanaści juz się
    #wysypuje xdddddddd, tu działa poprawnie i nie jest uznawany :(
    def solve(self, A, B):
        arr = list(A)
        
        counter = 0
        i = 0
        while i < len(arr):# tak jak w selection sort, ale tutaj maks B swapów
            maxIndex = i
            k = i+1
            #print("A[maxIndex]: ", arr[i], "arr: ", arr, "i: ", i)
            while k < len(arr):
                if arr[k] > arr[maxIndex]: # co z >=??
                    maxIndex = k
                if arr[k]==9:break
                k+=1
            
            if maxIndex!=i:#sprawdz, czy wykonac swap
                arr[i], arr[maxIndex] = arr[maxIndex], arr[i]
                counter +=1
            if counter ==B:break
            i +=1
        return "".join(arr)

File: Backtracking/CPP/MaximalString.cpp
void getmax(string A, int B, string& m, int &steps){
    if(B==0)
        return;
    for(int i=0;i<A.size()-1;i++)
        for(int j=i+1;j<A.size();j++){
            steps +=1;
            //cout<<"m: "<<m<<endl;
            if(A[j]>A[i]){
                //cout<<"innerIF"<<endl;
                swap(A[j], A[i]);
                if(A>m)
                    m = A;
                
                getmax(A, B-1, m, steps);
                swap(A[j], A[i]);
            }
        }
}
/*po prostu xd, tutaj program się wysypuje dla wartości 1034345678976
10 i uznaje poprawne rozwiązanie, a iteracyjna metoda, która jest szybsza i działa nawet dla 
większych przypadków uznawana jest jako błąd*/
string Solution::solve(string A, int B) {
    int steps = 0;
    string m = A;
    getmax(A, B, m, steps);
    //cout<<"steps: "<<steps<<endl;
    return m;
}